services/web/src/utils/middleware/history.js

```diff
@@ -0,0 +1,77 @@
+// Rewrites URLs for SPAs.
+//
+// Together with the template middleware this will allow SPAs
+// to serve templates for separate apps. It will match against
+// the base URLs of known apps and remove anything after. URLs
+// without a trailing slash will 301 redirect it appended.
+// Subdomains that match apps will also result in a rewrite.
+//
+// Usage:
+//
+// app.use(history({ for: ['/', '/admin/'] }))
+//
+// Examples:
+//
+// Domain          URL          Status  Result
+// ---------------------------------------------
+// www.site.com    /            200     /
+// www.site.com    /page        200     /
+// www.site.com    /admin/      200     /admin/
+// www.site.com    /admin/page  200     /admin/
+// www.site.com    /admin       301     /admin/
+// admin.site.com  /page        200     /admin/
+//
+
+const HOT_UPDATE_REG = /hot-update/;
+const EXT_URL = /\.[a-z]+$/i;
+
+function canRewriteUrl(path) {
+  return !hasFileExension(path) && !isHotUpdateUrl(path);
+}
+
+function hasFileExension(url) {
+  return EXT_URL.test(url);
+}
+
+function isHotUpdateUrl(url) {
+  return HOT_UPDATE_REG.test(url);
+}
+
+function matchSubdomainApp(subdomains, apps) {
+  if (subdomains.length === 1) {
+    const app = `/${subdomains[0]}/`;
+    return apps.find(a => a === app);
+  }
+  return null;
+}
+
+module.exports = function historyMiddleware(opt) {
```

Comment:
I like this personally.
Admin and App in my mind is 2 different services (granted there is an issue that needs to be solved around sharing components).  But i know @dominiek isnt seeing it the same way :) recall the talk about the docs.

For bare bedrock setup i wouldn't expect this level of complexity.  Generally speaking i think only big setups would require this. So i would say its overly complex for most of the setups

Reply:
Well this middleware is doing something fundamental for all apps, which is rewriting the URL for serving templates, which is something we've always had. What I added on top was actually pretty minimal which is the ability to specify more than one URL to rewrite to. So in most cases you would do this:

```
app.use(historyMiddleware({ apps: ['/'] }));
```

which is what it is by default, but you can optionally do this:

```
app.use(historyMiddleware({ apps: ['/' ,'/admin/'] }));
```

As to the question of whether or not admin should be a different service that's totally up to you. If you want a separate service then just leave everything here as is and copy the `web` directory I suppose (or however you want to set it up.